### ISSUE TITLE
Add a post about running Kythe on OS X. [phab:D293]

### ISSUE THREAD

**Issue by kythe** [google/kythe#512]
_Created by **zarko** at 2015-05-28 22:38:19_: 

If anyone hasn't built on OS X yet,
please try out the instructions and let me
know if I've left anything out.

**Comment by kythe** [comment]
_**bzz** wrote at 2015-05-29 11:26:41_:

not sure if that suppose to happen:
 XCode 6.1.1

`$sw_vers                                                                                             
ProductName:	Mac OS 
ProductVersion:	10.10.3
BuildVersion:	14#341
`
`$ bazel build //...                                                                               
INFO: Found 397 targets...
INFO: From Compiling third_party/rapidjson/test/unittest/valuetest.cc:
third_party/rapidjson/test/unittest/valuetest.cc:58:3: error: static_assert failed ""
  static_assert(!std::is_copy_assignable<Value>::value, "");
  ^             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
third_party/rapidjson/test/unittest/valuetest.cc:62:3: error: static_assert failed ""
  static_assert(!std::is_nothrow_copy_assignable<Value>::value, "");
  ^             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
2 errors generated.
ERROR: /Users/.../google-kythe-osx/third_party/rapidjson/BUILD:99:1: C++ compilation of rule '//third_party/rapidjson:rapidjson_value_test_lib' failed: clang failed: error executing command /Users/.../google-kythe-osx/tools/cpp/clang '-std=c++11' -fno-rtti -Wall -Werror -Wno-unused-variable -Wno-deprecated-declarations -D_XOPEN_SOURCE -static -Wno-non-virtual-dtor ... (remaining 25 argument(s) skipped): com.google.devtools.build.lib.shell.BadExitStatusException: Process exited with status 1.
INFO: Elapsed time: 26.369s, Critical Path: 21.09s`

**Comment by kythe** [comment]
_**schroederc** commented inline (context lost during migration) at 2015-05-29 15:20:40_:

Why not just
```
cd ~
git clone https://github.com/google/bazel/
cd bazel
```

**Comment by kythe** [comment]
_**schroederc** commented inline (context lost during migration) at 2015-05-29 15:20:40_:

Same as above

**Comment by kythe** [comment]
_**schroederc** commented inline (context lost during migration) at 2015-05-29 15:20:40_:

This can also be installed with `brew install go`.

**Comment by kythe** [comment]
_**schroederc** commented inline (context lost during migration) at 2015-05-29 15:20:40_:

Homebrew installation might be more useful before the Bazel install.  It simplified things for me greatly by installing xcode.

**Comment by kythe** [comment]
_**schroederc** wrote at 2015-05-29 15:20:40_:

I deployed this draft to https://79f82e731a5f3fe8dc17e7f3fded1898e4929f4e-dot-kythe-repo.appspot.com/2015/05/29/osx-support/

**Comment by kythe** [comment]
_**schroederc** commented inline (context lost during migration) at 2015-05-29 15:25:10_:

Go can also be installed with `brew install go`

**Comment by kythe** [comment]
_**schroederc** commented inline (context lost during migration) at 2015-05-29 15:26:55_:

Maybe point to http://kythe.io/phabricator/maniphest instead of just the Phabricator homepage?

**Comment by kythe** [comment]
_**schroederc** commented inline (context lost during migration) at 2015-05-29 15:26:55_:

kythe-dev instead of the community forum?

**Comment by kythe** [comment]
_**schroederc** wrote at 2015-05-29 15:57:22_:

`bazel test //...` works for me, but I'm getting a bunch of warnings.  Here's just a few.

```
INFO: From CcIndexer kythe/cxx/extractor/testdata/extract_verify_std_vector_test/compilation/kythe/cxx/extractor/testdata/extract_verify_std_vector_test.cc.entries:
Warning: when looking for /usr/include/pthread/sched.h[3efaeeac44fdd32a9c49dd5d779c71213013ed56571b03bc7587cd209d412505]:1022: missing source path
```

```
INFO: From Linking kythe/cxx/common/commandline_test:
ld: warning: option -noall_load is obsolete and being ignored
```

```
INFO: From Linking tools/cpp/libmalloc.a:
/Library/Developer/CommandLineTools/usr/bin/libtool: file: bazel-out/local_darwin-fastbuild/bin/tools/cpp/_objs/malloc/tools/cpp/empty.pic.o has no symbols
warning: /Library/Developer/CommandLineTools/usr/bin/libtool: warning for library: bazel-out/local_darwin-fastbuild/bin/tools/cpp/libmalloc.a the table of contents is empty (no object file members in the library define global symbols)
```

```
INFO: From Compiling third_party/grpc/src/core/tsi/ssl_transport_security.c:
third_party/grpc/src/core/tsi/ssl_transport_security.c:123:3: warning: 'SSL_library_init' is deprecated: first deprecated in OS X 10.7 [-Wdeprecated-declarations]
  SSL_library_init();
  ^
/usr/include/openssl/ssl.h:1569:5: note: 'SSL_library_init' has been explicitly marked deprecated here
int SSL_library_init(void ) DEPRECATED_IN_MAC_OS_X_VERSION_10_7_AND_LATER;
    ^
third_party/grpc/src/core/tsi/ssl_transport_security.c:124:3: warning: 'SSL_load_error_strings' is deprecated: first deprecated in OS X 10.7 [-Wdeprecated-declarations]
  SSL_load_error_strings();
  ^
/usr/include/openssl/ssl.h:1432:6: note: 'SSL_load_error_strings' has been explicitly marked deprecated here
void    SSL_load_error_strings(void ) DEPRECATED_IN_MAC_OS_X_VERSION_10_7_AND_LATER;
        ^
third_party/grpc/src/core/tsi/ssl_transport_security.c:125:3: warning: 'OPENSSL_add_all_algorithms_noconf' is deprecated: first deprecated in OS X 10.7 [-Wdeprecated-declarations]
  OpenSSL_add_all_algorithms();
  ^
/usr/include/openssl/evp.h:829:3: note: expanded from macro 'OpenSSL_add_all_algorithms'
                OPENSSL_add_all_algorithms_noconf()
                ^
```

```
INFO: From Linking third_party/grpc/libgrpc_unsecure.a:
/Library/Developer/CommandLineTools/usr/bin/libtool: file: bazel-out/local_darwin-fastbuild/bin/third_party/grpc/_objs/grpc_unsecure/third_party/grpc/src/core/iomgr/iocp_windows.pic.o has no symbols
/Library/Developer/CommandLineTools/usr/bin/libtool: file: bazel-out/local_darwin-fastbuild/bin/third_party/grpc/_objs/grpc_unsecure/third_party/grpc/src/core/iomgr/iomgr_windows.pic.o has no symbols
/Library/Developer/CommandLineTools/usr/bin/libtool: file: bazel-out/local_darwin-fastbuild/bin/third_party/grpc/_objs/grpc_unsecure/third_party/grpc/src/core/iomgr/pollset_multipoller_with_epoll.pic.o has no symbols
/Library/Developer/CommandLineTools/usr/bin/libtool: file: bazel-out/local_darwin-fastbuild/bin/third_party/grpc/_objs/grpc_unsecure/third_party/grpc/src/core/iomgr/pollset_windows.pic.o has no symbols
/Library/Developer/CommandLineTools/usr/bin/libtool: file: bazel-out/local_darwin-fastbuild/bin/third_party/grpc/_objs/grpc_unsecure/third_party/grpc/src/core/iomgr/resolve_address_windows.pic.o has no symbols
/Library/Developer/CommandLineTools/usr/bin/libtool: file: bazel-out/local_darwin-fastbuild/bin/third_party/grpc/_objs/grpc_unsecure/third_party/grpc/src/core/iomgr/socket_utils_linux.pic.o has no symbols
/Library/Developer/CommandLineTools/usr/bin/libtool: file: bazel-out/local_darwin-fastbuild/bin/third_party/grpc/_objs/grpc_unsecure/third_party/grpc/src/core/iomgr/socket_windows.pic.o has no symbols
/Library/Developer/CommandLineTools/usr/bin/libtool: file: bazel-out/local_darwin-fastbuild/bin/third_party/grpc/_objs/grpc_unsecure/third_party/grpc/src/core/iomgr/tcp_client_windows.pic.o has no symbols
/Library/Developer/CommandLineTools/usr/bin/libtool: file: bazel-out/local_darwin-fastbuild/bin/third_party/grpc/_objs/grpc_unsecure/third_party/grpc/src/core/iomgr/tcp_server_windows.pic.o has no symbols
/Library/Developer/CommandLineTools/usr/bin/libtool: file: bazel-out/local_darwin-fastbuild/bin/third_party/grpc/_objs/grpc_unsecure/third_party/grpc/src/core/iomgr/tcp_windows.pic.o has no symbols
/Library/Developer/CommandLineTools/usr/bin/libtool: file: bazel-out/local_darwin-fastbuild/bin/third_party/grpc/_objs/grpc_unsecure/third_party/grpc/src/core/iomgr/wakeup_fd_eventfd.pic.o has no symbols
```

**Comment by kythe** [comment]
_**schroederc** wrote at 2015-05-29 16:00:42_:

There seems to be an error ignored while building the docs:

```
INFO: From Executing genrule //kythe/docs:kythe-storage:
/bin/sh: : command not found
asciidoc: WARNING: kythe-storage.txt: line 106: filter non-zero exit code: "" "/usr/local/Cellar/asciidoc/8.6.9/etc/asciidoc/filters/graphviz/graphviz2png.py"  -o "/private/var/tmp/_bazel_schroederc/74c4715b4953fb15a59589ef48454c24/kythe/bazel-out/local_darwin-fastbuild/bin/kythe/docs/kythe-storage__1.png" -L dot -F png -: returned 127
asciidoc: WARNING: kythe-storage.txt: line 106: no output from filter: "" "/usr/local/Cellar/asciidoc/8.6.9/etc/asciidoc/filters/graphviz/graphviz2png.py"  -o "/private/var/tmp/_bazel_schroederc/74c4715b4953fb15a59589ef48454c24/kythe/bazel-out/local_darwin-fastbuild/bin/kythe/docs/kythe-storage__1.png" -L dot -F png -
```

**Comment by kythe** [comment]
_**schroederc** wrote at 2015-05-29 16:04:58_:

Building ``//kythe/docs/schema`` is failing as well.  It emits messages like this for a few pages.

```
INFO: From Executing genrule //kythe/docs/schema:schema:
usage: mktemp [-d] [-q] [-t prefix] [-u] template ...
       mktemp [-d] [-q] [-u] -t prefix 
mkdir: /example: Permission denied
asciidoc: WARNING: schema.txt: line 75: filter non-zero exit code: "kythe/docs/schema/example.sh" xhtml11 kythe C++ "Typedefs are aliases.": returned 1
asciidoc: WARNING: schema.txt: line 75: no output from filter: "kythe/docs/schema/example.sh" xhtml11 kythe C++ "Typedefs are aliases."
usage: mktemp [-d] [-q] [-t prefix] [-u] template ...
       mktemp [-d] [-q] [-u] -t prefix 
/private/var/tmp/_bazel_schroederc/74c4715b4953fb15a59589ef48454c24/kythe/kythe/docs/schema/example-java.sh: line 35: /E.java: Permission denied
asciidoc: WARNING: schema.txt: line 96: filter non-zero exit code: "kythe/docs/schema/example.sh" xhtml11 kythe Java "Classes can be annotated.": returned 1
asciidoc: WARNING: schema.txt: line 96: no output from filter: "kythe/docs/schema/example.sh" xhtml11 kythe Java "Classes can be annotated."
usage: mktemp [-d] [-q] [-t prefix] [-u] template ...
       mktemp [-d] [-q] [-u] -t prefix 
mkdir: /example: Permission denied
```

**Comment by kythe** [comment]
_**zarko** wrote at 2015-05-29 20:42:19_:

Responding in order from top to bottom to the out-of-line comments:

bzz: That is weird. Do you know what version of libc++ you're using?
If you remove the tests from //third_party/rapidjson/BUILD (all the targets
beyond and including "rapidjson_unit_test_lib") can you get any further?

schroederc: I don't see anything curious in the linker and deprecation warnings.

I hadn't noticed the schema problem (I think we should remove the
manual tag from the BUILD file--it's supposed to be run as a test).
#518 should help.

The "missing source path" warning is interesting. I'd seen it before
but failed to create #515. It relates to claiming and can result in potentially
more redundant entries being emitted.

Re: no output from filter: ""
  "/usr/local/Cellar/asciidoc/8.6.9/etc/asciidoc/filters/graphviz/graphviz2png.py",
it sounds like something's wrong with your asciidoc installation--it's trying to
execute the empty string applied to a Python script.

**Comment by kythe** [comment]
_**zarko** wrote at 2015-05-29 20:57:05_:

I'm seeing the same thing now re: //kythe/docs:kythe-storage (if only asciidoc failed if a filter failed).

**Comment by kythe** [comment]
_**zarko** wrote at 2015-05-29 22:18:23_:

  - Address comments and fix asciidoc invocation

**Comment by kythe** [comment]
_**bzz** wrote at 2015-05-30 01:55:05_:

zarko: indeed after commenting out rapidjson tests and pulling latest changes was able to move further

1. now build fails with
```ERROR: missing input file '//tools/go:go'.
ERROR: /Users/.../google-kythe-osx/third_party/grpc/BUILD:121:1: C++ compilation of rule '//third_party/grpc:grpc' failed: clang failed: error executing command /Users/.../google-kythe-osx/tools/cpp/clang -Wno-implicit-function-declaration -iquote . -iquote bazel-out/local_darwin-fastbuild/genfiles -isystem third_party/grpc/include -isystem ... (remaining 16 argument(s) skipped): com.google.devtools.build.lib.shell.AbnormalTerminationException: Process terminated by signal 15.
ERROR: /Users/.../google-kythe-osx/kythe/go/platform/tools/BUILD:5:1: //kythe/go/platform/tools:entrystream: missing input file '//tools/go:go'.
ERROR: /Users/.../google-kythe-osx/kythe/go/platform/tools/BUILD:5:1 1 input file(s) do not exist.
```
have go installed using brew
```$ ls -la tools/go                                                                                 
lrwxr-xr-x  1 xxxx   25 May 30 01:02 go -> ../Cellar/go/1.4.1/bin/go```

2. there are plenty of warnings like
```
INFO: From Linking third_party/liblibcurl.a:
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/libtool: file: bazel-out/local_darwin-fastbuild/bin/third_party/_objs/libcurl/third_party/empty.pic.o has no symbols
warning: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/libtool: warning for library: bazel-out/local_darwin-fastbuild/bin/third_party/liblibcurl.a the table of contents is empty (no object file members in the library define global symbols)
INFO: From Linking tools/cpp/libmalloc.a:
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/libtool: file: bazel-out/local_darwin-fastbuild/bin/tools/cpp/_objs/malloc/tools/cpp/empty.pic.o has no symbols
warning: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/libtool: warning for library: bazel-out/local_darwin-fastbuild/bin/tools/cpp/libmalloc.a the table of contents is empty (no object file members in the library define global symbols)
```

I'm not sure about version of libc++ being used, there are `/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1` and `/usr/local/Cellar/boost/1.58.0/include/boost/tr1/tr1/` and `/usr/include/c++/4.2.1/`.
Would appreciate direction on how to check that.

**Comment by kythe** [comment]
_**bzz** wrote at 2015-05-30 12:46:35_:

`bazel build //...` passes after manually symlinking `tools/go/go` -> `/usr/local/bin/go` and removing all test targets from `//third_party/rapidjson/BUILD`

with deprecation warnings 

```/usr/include/openssl/ssl.h:1514:6: note: 'SSL_CTX_ctrl' has been explicitly marked deprecated here
long    SSL_CTX_ctrl(SSL_CTX *ctx,int cmd, long larg, void *parg) DEPRECATED_IN_MAC_OS_X_VERSION_10_7_AND_LATER;
        ^
108 warnings generated.```

and a lot of things like

```warning: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/libtool: warning for library: bazel-out/local_darwin-fastbuild/bin/third_party/liblibuuid.a the table of contents is empty (no object file members in the library define global symbols)```

Also, pulling the fix for asciidoc invocation resolves the issue with `//kythe/docs/schema`


`bazel test //...` works but has some tests failing

```
INFO: Found 303 targets and 72 test targets...

//kythe/cxx/common/testdata:net_client_test_runner                       FAILED
//kythe/cxx/tools/fyi/testdata:basic                                     FAILED
//kythe/cxx/tools/fyi/testdata:hopeless                                  FAILED
//kythe/cxx/tools/fyi/testdata:nothing                                   FAILED
//kythe/go/serving/tools/testdata:kwazthis_test                          FAILED

Executed 72 out of 72 tests: 67 tests pass and 5 fail locally.

```

**Comment by kythe** [comment]
_**zarko** wrote at 2015-05-30 22:43:56_:

bzz: Did you see any errors when you ran setup_bazel.sh? That should have created the go symlink. (Maybe we're seeing the same kind of thing that we saw in the asciidoc path and we need to look for tools on OS X using `command -v`.)

Could you run `//kythe/cxx/common/testdata:net_client_test_runner` alone (`bazel test //kythe/cxx/common/testdata:net_client_test_runner`) and post or e-mail me the test log directly? (It's just my Phabricator username `google.com`). The other tests that are failing for you depend on that one succeeding; `net_client_test_runner` tests the underlying infrastructure that those other tests use to spin up an xrefs service.

I'm assuming that your libc++ version is similar to mine (`v1`).

**Comment by kythe** [comment]
_**bzz** wrote at 2015-05-31 05:18:01_:

zarko there were no errors on `setup_bazel.sh`, but #514 might have already took care of it.

Logs of `bazel test //kythe/cxx/common/testdata:net_client_test_runner` are here https://gist.github.com/bzz/0081a7e78a5356f2fb55

**Comment by kythe** [comment]
_**zarko** wrote at 2015-06-01 17:39:40_:

If you change the line that reads

`if [[ -e /proc/1/cgroup ]] && grep -qe ':/docker' /proc/1/cgroup ; then`

to

`if true ; then`

in `kythe/cxx/common/testdata/start_http_service.sh`, do you get any farther?

**Comment by kythe** [comment]
_**schroederc** commented inline (context lost during migration) at 2015-06-01 23:10:28_:

This file will need to be renamed with a new date once this change is accepted.

**Comment by kythe** [comment]
_**schroederc** required changes at 2015-06-01 23:10:28_

**Comment by kythe** [comment]
_**bzz** wrote at 2015-06-03 01:20:54_:

Changing a line in `kythe/cxx/common/testdata/start_http_service.sh` as suggested got me further, but not much - still the same tests are failing, here are new logs 

```
$less /private/var/tmp/_bazel_bzz/3d1bfd70d37a0aadb174e62056949bf3/google-kythe-osx/bazel-out/local_darwin-fastbuild/testlogs/kythe/cxx/common/testdata/net_client_test_runner/test.log
exec ${PAGER:-/usr/bin/less} "$0" || exit 1
kythe/cxx/common/testdata/start_http_service.sh: line 63: /proc/sys/net/ipv4/ip_local_port_range: No such file or directory
```

**Comment by kythe** [comment]
_**zarko** wrote at 2015-06-04 21:36:10_:

  - Update date and rebase.

**Comment by kythe** [comment]
_**schroederc** commented inline (context lost during migration) at 2015-06-04 22:28:43_:

Link Bazel to bazel.io

**Comment by kythe** [comment]
_**schroederc** commented inline (context lost during migration) at 2015-06-04 22:28:43_:

No need to assume where they clone the repository.

**Comment by kythe** [comment]
_**zarko** wrote at 2015-06-04 22:44:27_:

  - Link to Bazel and remove a cd.

**Comment by kythe** [comment]
_**schroederc** accepted this revision at 2015-06-04 22:47:11_

**Comment by kythe** [comment]
_**zarko** closed via commit 9164bc3357bad31bba5eab89d23f56e9890c4dff at 2015-06-05 18:12:43_
